File: TeAchei/src/main/java/com/teachei/api/application/usecase/AlterarSenhaUseCaseImpl.java
package com.teachei.api.application.usecase;

import com.teachei.api.application.ports.in.AlterarSenhaUseCase;
import com.teachei.api.shared.security.PasswordEncoderPort;
import com.teachei.api.application.ports.out.UsuarioRepositoryPort;
import com.teachei.api.domain.exception.CredenciaisInvalidasException;
import com.teachei.api.domain.exception.UsuarioNaoEncontradoException;
import com.teachei.api.domain.model.Usuario;

import java.util.UUID;

/**
 * Implementation of the password change use case.
 */
public class AlterarSenhaUseCaseImpl implements AlterarSenhaUseCase {

    private final UsuarioRepositoryPort usuarioRepository;
    private final PasswordEncoderPort passwordEncoder;

    public AlterarSenhaUseCaseImpl(UsuarioRepositoryPort usuarioRepository,
                                    PasswordEncoderPort passwordEncoder) {
        this.usuarioRepository = usuarioRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @Override
    public void executar(UUID usuarioId, AlterarSenhaCommand command) {
        // Find user
        Usuario usuario = usuarioRepository.buscarPorId(usuarioId)
            .orElseThrow(() -> new UsuarioNaoEncontradoException(usuarioId));

        // Verify current password
        if (!passwordEncoder.matches(command.senhaAtual(), usuario.getSenha())) {
            throw new CredenciaisInvalidasException("Senha atual incorreta");
        }

        // Encode new password
        String novaSenhaHash = passwordEncoder.encode(command.novaSenha());

        // Update password
        usuario.alterarSenha(novaSenhaHash);

        // Save
        usuarioRepository.salvar(usuario);
    }
}

File: teachei-web/ap/auth/application/usecase/AlterarSenhaUseCaseImpl.ts
import bcrypt from "bcryptjs";
import { CredenciaisInvalidasException } from "@/ap/auth/domain/exception/CredenciaisInvalidasException";
import { NotFoundError } from "@/ap/shared/errors";
import type { AlterarSenhaUseCase } from "@/ap/auth/application/ports/in/AlterarSenhaUseCase";
import type { UsuarioRepositoryPort } from "@/ap/auth/application/ports/out/UsuarioRepositoryPort";

export class AlterarSenhaUseCaseImpl implements AlterarSenhaUseCase {
  constructor(private repo: UsuarioRepositoryPort) {}

  async execute(usuarioId: string, senhaAtual: string, novaSenha: string): Promise<void> {
    const usuario = await this.repo.findById(usuarioId);
    if (!usuario) throw new NotFoundError("Usuário não encontrado");
    if (!usuario.senhaHash) throw new CredenciaisInvalidasException();

    const valid = await bcrypt.compare(senhaAtual, usuario.senhaHash);
    if (!valid) throw new CredenciaisInvalidasException();

    const novaSenhaHash = await bcrypt.hash(novaSenha, 10);
    await this.repo.updateSenha(usuarioId, novaSenhaHash);
  }
}
